src/main/java/fr/xephi/authme/service/SpectateLoginService.java

```diff
@@ -0,0 +1,86 @@
+package fr.xephi.authme.service;
+
+import fr.xephi.authme.settings.properties.RestrictionSettings;
+import org.bukkit.GameMode;
+import org.bukkit.Location;
+import org.bukkit.entity.ArmorStand;
+import org.bukkit.entity.Player;
+
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SpectateLoginService {
+
+    private Map<Player, ArmorStand> armorStands = new HashMap<>();
+    private Map<Player, GameMode> gameModeMap = new HashMap<>();
+
+    @Inject
+    private CommonService service;
+
+    /**
+     * Creates a stand for the player
+     *
+     * @param player the player
+     */
+    public void createStand(Player player) {
+        Location location = player.getLocation();
+        ArmorStand stand = spawnStand(location);
+
+        armorStands.put(player, stand);
+        gameModeMap.put(player, player.getGameMode());
```

Comment:
What if the server shuts down while players are not logged in? I guess they'll be saved with the spectator game mode, and next time when they'll join, they'll already have spectator game mode and won't be able to leave it. I think this is quite relevant because it's not only an inconvenience to a player, it's giving them access to a game mode that many servers don't want to give to players... 
I could imagine trying to make use of the limbo player stuff or to maybe have an additional setting that will set players to survival instead of spectator when the code would want to. It's ugly but we've had to add a few of those failsafe properties in the past and people seem happy with it

Reply:
Apparently, when the plugin disabled, I can call a function that will remove all armorstands and return the players to their previous state. Is it possible to implement it this way?

Reply:
Could you please tell me if I can call SpectateLoginService in onDisable directly to remove armorstands, or is this not a good practice?

Reply:
Sorry for the late reply.
I don't see any better way. I think you could get the SpecateLoginService with `injector.getIfAvailable(SpectateLoginService.class)` in AuthMe#onDisable. But still, when a server unexpectedly shuts down (if it crashes or its process is terminated abruptly) we have the problem that players are potentially in spectator mode, and there are armor stands that no one tracks anymore. I don't know if the armor stands are an issue (I have zero experience with the way you spawn them). 
Wondering if we should have a property that specifies what game mode a player should change to if after authentication, we would put him into spectate game mode and/or to add the game mode to the LimboPlayer. I can elaborate on the whole limbo player thing if it helps

Reply:
Untraceable armorstand is not as critical as players in Spectator gamemode. It would be really cool if it could be done with LimboPlayer

Reply:
What about adding a "limboArmorStandUUID" field to LimboPlayer?

Reply:
Do you mean adding a way to keep track of whether a player is assigned an armorstand?

Reply:
@mrchefran yes, exactly